### PR TITLE
[Transformations] Tolerate FakeQuantize on KV-cache concat in StateManagementPattern

### PR DESCRIPTION
**Pull request by evkotov**
### Details

Full int8 activation quantization inserts a FakeQuantize on the concatenated KV-cache activation, directly after the KV-cache Concat. This happens when a model is exported with optimum-cli export openvino --quant-mode int8 --dataset <ds> (a8w8 / SmoothQuant, weights and activations). StateManagementPattern matched the K/V paths straight from the Concat with no tolerance for an intervening FakeQuantize, so SDPA was never rewritten to PagedAttention. The beam_idx and attention_mask consumer subgraphs then survived, and the final validate_nodes_and_infer_types() in SDPAToPagedAttention failed with:

Model references undeclared parameters:
opset1::Parameter beam_idx () -> (i32[?])
opset1::Parameter attention_mask () -> (i64[?,?])
Weight-only int8 (optimum-cli export openvino --weight-format int8) quantizes only weights (no FakeQuantize on activations) and is unaffected.

This wraps k_concat/v_concat in pattern::optional<v0::FakeQuantize> right after the Concat, before any downstream KV shaping. This is the actual NNCF insertion point and covers both the non-GQA case (FQ directly on the SDPA input) and the GQA/MQA case (FQ before the Unsqueeze-Broadcast-Reshape repeat_kv expansion).

The callback rebuilds the K/V PagedAttention inputs from the pre-concat "current" K/V tensors, which bypasses this FakeQuantize, so the matched FakeQuantize is re-applied onto the rebuilt K/V feed — mirroring the FakeConvert handling for FP8. The re-apply runs just before the FakeConvert loop, so when both are present the upstream concat FakeQuantize stays inner and the SDPA-input FakeConvert outer, matching the original graph order. Only per-tensor limits (what a8w8 produces) are re-applied. Preserving the quantization keeps the PagedAttention path numerically consistent with the model's calibrated int8 activations (and with the stateful SDPA path); dropping it would leave K/V at full precision while Q kept its own FakeQuantize.

### Tests

sdpa_to_paged_attention_test.cpp: added a parameterized test over {fq_on_k, fq_on_v, gqa} (8 combinations, GQA and non-GQA). Each builds a stateful single-layer SDPA model, optionally inserts a FakeQuantize on the KV path, runs SDPAToPagedAttention, and asserts the conversion succeeds (exactly one PagedAttentionExtension, beam_idx/attention_mask removed) and that the FakeQuantize is preserved on the corresponding PagedAttention K/V input.

Verified end-to-end on meta-llama/Llama-3.2-1B-Instruct exported with optimum-cli export openvino --quant-mode int8 --dataset wikitext2, loaded through the continuous-batching path.

### Tickets:
 - 182239

### CODE REVIEW

**Comment by CuriousPanCake** [comment]
Should I extend my PR https://github.com/openvinotoolkit/openvino/pull/36299 with your changes?

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 1]
```diff
@@ -302,6 +303,12 @@ static std::shared_ptr<ov::Node> optional_fake_convert(const std::shared_ptr<ov:
     return std::make_shared<Or>(OutputVector{fc_2, fc_3, input});
 }
 
+static std::shared_ptr<ov::Node> optional_fake_quantize(const std::shared_ptr<ov::Node>& input) {
+    // v0::FakeQuantize always has 5 inputs; pattern::optional matches multi-input ops by arity and
+    // does not synthesize missing ones, so all 5 inputs must be listed.
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]:
```suggestion
```

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 2]
```diff
@@ -355,6 +362,17 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
...
 
+    // Full int8 activation quantization (a8w8 / SmoothQuant) inserts a FakeQuantize on the
+    // concatenated KV-cache activation, i.e. directly after this Concat and before whatever consumes
+    // it: the SDPA K/V input (non-GQA) or the Unsqueeze-Broadcast-Reshape "repeat_kv" expansion
+    // (GQA/MQA). Tolerate an optional FakeQuantize here so every downstream KV path (kv_shaping,
+    // *_simply_shaped, *_shaped_transposed and the direct-to-SDPA Or) still binds. The dequant
+    // FakeQuantize is not needed on the PagedAttention KV path (which is rebuilt from the pre-concat
+    // "current" K/V tensors), so it is simply dropped. Unlike the FakeConvert case above (FP8, kept
+    // on the SDPA inputs), this FakeQuantize is not re-applied in the callback.
+    k_concat = optional_fake_quantize(k_concat);
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Nothing to do in callback?

Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Probably no need for the helper since only one version of FakeQuantize exists

Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: You didn't reply to my question.

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 3]
```diff
@@ -355,6 +356,17 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
...
     v_concat = std::make_shared<Or>(OutputVector{kv_concat_split->output(1), v_concat});
 
+    // Full int8 activation quantization (a8w8 / SmoothQuant) inserts a FakeQuantize on the
+    // concatenated KV-cache activation, i.e. directly after this Concat and before whatever consumes
+    // it: the SDPA K/V input (non-GQA) or the Unsqueeze-Broadcast-Reshape "repeat_kv" expansion
+    // (GQA/MQA). Tolerate an optional FakeQuantize here so every downstream KV path (kv_shaping,
+    // *_simply_shaped, *_shaped_transposed and the direct-to-SDPA Or) still binds. The dequant
+    // FakeQuantize is not needed on the PagedAttention KV path (which is rebuilt from the pre-concat
+    // "current" K/V tensors), so it is simply dropped. Unlike the FakeConvert case above (FP8, kept
+    // on the SDPA inputs), this FakeQuantize is not re-applied in the callback.
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: > Unlike the FakeConvert case above (FP8, kept on the SDPA inputs), this FakeQuantize is not re-applied in the callback.

Why? If they're in the graph, they're for a reason here.

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp (hunk 2 above)
Comment (evkotov) [reply src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Originally I dropped the FakeQuantize here, but that was wrong: the callback rebuilds the K/V PagedAttention inputs from the pre-concat "current" tensors (via take_4d), which bypasses this concat-level FakeQuantize, so the activation quantization was silently lost on the PA path.

Fixed in e231e3058d: the matched FakeQuantize is now re-applied onto the rebuilt K/V feed in the callback, right next to the FakeConvert re-apply loop (it runs first, so when both are present the upstream concat FQ stays inner and the SDPA-input FakeConvert outer, matching the original graph order). Guarded to per-tensor limits, which is what a8w8/SmoothQuant produces.

It's not cosmetic — dropping vs keeping it changes the logits by ~11% mean (max 1.76) and flips ~25% of the greedy argmax tokens on Llama-3.2-1B-Instruct a8w8, and with it preserved the continuous-batching (PA) path now matches the stateful SDPA path token-for-token. The regression test was updated to assert the FakeQuantize survives on the PA K/V input instead of being removed.

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp (hunk 3 above)
Comment (evkotov) [reply src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Dropping it was wrong: the FakeQuantize is there for a reason (a8w8 / SmoothQuant activation quantization on the K/V cache), and dropping it lost that quantization on the PagedAttention path while Q kept its own FakeQuantize — so the PA path was computing Q(int8)·K(fp32), which isn't the model's calibrated behavior.

Fixed in e231e3058d: the matched FakeQuantize is now re-applied onto the rebuilt K/V PA feed in the callback, exactly like the FakeConvert. It runs just before the FakeConvert loop, so when both are present the upstream concat FQ nests inner and the SDPA-input FakeConvert outer, matching the original graph order. Guarded to per-tensor limits (what a8w8 produces).

The comment you quoted has been rewritten accordingly ("...the matched FakeQuantize is preserved: re-applied onto the rebuilt K/V PagedAttention feed"). Impact isn't cosmetic — on Llama-3.2-1B-Instruct a8w8, drop-vs-keep changes the logits by ~11% mean (max 1.76) and flips ~25% of greedy argmax tokens; with it preserved, the continuous-batching path now matches the stateful SDPA path token-for-token. The regression test now asserts the FakeQuantize survives on the PA K/V input instead of being removed.

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 4]
```diff
@@ -355,6 +356,28 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
...
+        if (!fq)
+            return false;
+        for (size_t i = 1; i < fq->get_input_size(); ++i) {
+            const auto& ps = fq->get_input_partial_shape(i);
+            if (!ps.is_static() || ov::shape_size(ps.to_shape()) != 1)
+                return false;
+        }
+        return true;
+    };
+    k_concat = pattern::optional<v0::FakeQuantize>({k_concat, any_input(), any_input(), any_input(), any_input()},
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Looks like we can extend the optional_fake_convert(), eg optional_quantization() and add this pattern there

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 5]
```diff
@@ -562,10 +585,26 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
...
-        // Re-apply FakeConvert after Reshape for Q/K/V (aligned position for all three)
         Output<Node> q_to_pa = q_reshape;
         Output<Node> k_to_pa = k_reshape;
         Output<Node> v_to_pa = v_reshape;
+
+        // Re-apply the KV-cache FakeQuantize (matched at k_concat/v_concat) onto the rebuilt PA feed;
+        // runs before the FakeConvert loop to keep the original op order when both are present. The
+        // pattern only binds a per-tensor FakeQuantize, so its scalar limits broadcast onto the
+        // flattened feed and re-applying it is exact.
+        for (auto& [pattern_node, target] : {std::tie(k_concat, k_to_pa), std::tie(v_concat, v_to_pa)}) {
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: that the same re-applying logic as already done for fakeConvert, can we reuse it?

**Review by mryzhov — APPROVED** [review]
